### DISCUSSION TITLE
Moving the GPS button below the compass

### DISCUSSION THREAD

**Discussion by smichel17** [streetcomplete/StreetComplete#3326] [Ideas]
This is kind of a follow-up to https://github.com/streetcomplete/StreetComplete/issues/2765#issuecomment-927184410

I'm wondering if simply putting the gps button near the compass would be enough of a hint that the two are related. We could even make the gps button hide itself by animating up "behind" the compass button, to hint that the functionality has changed.

**Comment by smichel17** [comment]
Some people who might be interested: @mnalis @Jonisos @ChildishGiant

(If anyone else wants to round up more people in past threads who have been confused by this, I'd appreciate it!)

**Comment by mnalis** [comment]
Well, after months of training following your [great explanation](https://wiki.openstreetmap.org/wiki/StreetComplete/FAQ#How_to_use_2D.2F3D_and_GPS_following.2Frotation.3F) (which I'm keeping on the phone for reference), I'm _almost_ getting a hang of how it works in _most cases_. At least, I now always know how to return to known usable state of north-is-up-2D-mode :smile:  

Still, I'm heavily tempted to one day add `2D only` preference in my fork and make each button do exactly one thing.  Such overloading of UI elements (so they do completely different thing depending on some previous state and in which order you press them) is highly unintuitive and taxing for my poor brain.

I'm not sure putting the icons close to each other would do anything for me (and I would actually advise against it, see muscle-memory issue in [another discussion](https://github.com/streetcomplete/StreetComplete/discussions/3330#discussioncomment-1399033)). 

Auto-hiding GPS button behind it might be helping **a little** but mostly by reducing a number of elements one need to guess about what they do if pressed (as it does already) than by the specific place it disappears to. Buy maybe it's just me now (after debugging that note location issue) being totally desensitized about where the icon dissapears to, as the note icons disappear to seemingly totally random places. It ***shouldn't hurt***, at least (**if the GPS icon remained where it is** at the bottom, and then in disappearing act take all that trip to the top. Might it make more visible, even).

What would help me is adding _another_ button (or preference checkbox, or menu item), used exclusively to `switch 2D/3D` view. Then each action could do one thing only, and thus IMHO be **much more intuitive**/understandable:
- pressing `2D/3D` icon would switch between 2D/birdseye view and tilted 3D view, while leaving other features as-is (or, having `2d/3d` as a preference in the options, where the user must explicitly choose which one they want to use might save screen estate - and would incidentally allow for fixing [#2206](https://github.com/streetcomplete/StreetComplete/issues/2206) too). 
  Alternatively, making entering `3D` mode possible only by two-finger swipe (or, for discoverability, as a menu item in `multi-function` icon, same as [add note](https://github.com/streetcomplete/StreetComplete/discussions/3330#discussioncomment-1399033) idea). In any case, removing it as additionally overloaded functionality on existing `compass` icon. 
- pressing `compass` icon would switch between `north is up` and `follow orientation` (and third option, unselectable by pressing the icon: `north is elsewhere (manual rotation)` which would point to where north is when user manually rotates the map)
- pressing `GPS dot` icon would zoom in and lock to your location (and disappear), as it does already.
- moving or rotating a screen manually would disengage both the `follow location` and `follow orientation` (and thus show `GPS dot` icon again, and make `compass` icon to assume `north is elsewhere (manual rotation)` look).

**Comment by smichel17** [comment]
> Such overloading of UI elements

*In general*, I agree with you that it's preferable not to overload UI elements. Also, I'd rather not split functionality unnecessarily; ideally the compass would swap only between the different rotation modes, and _no other elements would affect the rotation_. In this particular case, though, it's a trade-off against very limited screen space, and some of the combinations of options not making any sense.

- "Follow compass/[movement](https://github.com/streetcomplete/StreetComplete/issues/1040) direction" makes no sense with an unlocked camera, because there's no sensible point to auto-rotate around. Plus, the user will need to manually scroll the map anyway, at which point you might as well manually rotate, too.

- Locked + tilted + do not follow compass is going to make the map point the wrong direction as soon as you make a turn. It doesn't make sense without some way to rotate without unlocking ([final bullet point](https://github.com/streetcomplete/StreetComplete/issues/2191#issue-728302440)). But this would involve introducing a new type of locked mode: unlocked, locked, and "very locked", where scrolling rotates the map instead of unlocking.

So, either we split up the functionality between multiple independent buttons, using more screen real estate and allowing nonsensical states, or we overload one of the buttons. I don't think it's too much of a stretch to use the compass for determining both bearing and tilt, if you think of the compass less as a literal compass and more as a pointer which shows the orientation that the camera is facing.

**Reply by mnalis** [reply]
Good point about nonsensical states (although that maybe might be worked around by hiding buttons sometimes, as we do for GPS button at the moment). But that particular overloading is still confusing. Perhaps what could be done with minimal work to improve situation, in combination with [converting upload button to multi-function button](https://github.com/streetcomplete/StreetComplete/discussions/3330#discussioncomment-1399033):


- remove 2D/3D overloading from compass button (that would remove much of the UI confusion!)
- add `switch 3D mode` as an item to   `multi-function` button (that would restore, and even improve, on discoverability of the 3D feature which we removed in previous step), and make clicking it enable `3D view tilt` + `follow compass direction` + `locked camera`
- leave two-finger swipe as it is as quick alternate method to enter/leave 3D mode for more experienced users

**Reply by smichel17** [reply]
I don't think I like moving `switch 3D mode` into a multi-function menu. At least for me, changing the view of the map needs to have very responsive controls or it *feels* bad to use (in other words: opening a menu first is out). Also, based on that thread, I might end up not proposing a multi-function menu, anyway, since I am likely going to propose [removing the undo menu](https://github.com/streetcomplete/StreetComplete/discussions/3330#discussioncomment-1399451).

**Reply by mnalis** [reply]
Sure, I'm just giving some ideas that come to my mind. I rarely use 3D mode though, so that surely is something to be aware of.

**Comment by mnalis** [comment]
Or, perhaps, replace both `compass` and `GPS` icons with just one `map mode` icon, which would cycle through small number of actually useful modes on each button click (and change how that icon looks):

- Full 3D mode (`3D view tilt` + `follow compass direction` + `locked camera`)
- Full 2D mode (`2D/no tilt` + `follow compass direction` + `locked camera`)
- 2D no rotate (`2D/no tilt` + `locked camera`, and disable `follow compass direction`)
- 2D North is up (`2D/no tilt` + `North is up` + `locked camera`, and disable `follow compass direction`)

There would also be `fully manual` mode (in which icon would change to show where is North, like it does currently), and which would be activated not by clicking on that icon, but by user manually panning or rotating the map, and which would automatically disengage both `follow compass direction` and `locked camera` (but not modify `tilt` state).

When `map mode` icon is pressed after it has been put in `fully manual` mode (by panning/rotating map manually), it should first return to last mode it was in (for example if before panning it was in `2D North is up` mode, first click on it should return to that same  `2D North is up` mode, while further presses would follow normal cycling as always - next press would go to `Full 3D` in this example etc.)

That would accomplish:
- reduce screen estate used to just one button (instead of current two, or suggested three)
- reduce confusion greatly, as there is now just one button which does simple cycling ( `A` -> `B` -> `C` -> `D` ->  `A` -> `...`) instead of multiple buttons which depend on previous states and/or order in which they are pressed
- retain most (all?) of the usable map modes (or did I miss some?)
- greatly improve discoverability of the modes (as there is now just one button to cycle through all of modes, and not some sort of hard-to-discover [Mortal Kombat](https://www.mksecrets.net/mk1/eng/mk-specialmoves.php) fatality [combos](https://en.wikipedia.org/wiki/Combo_(video_games)), there is much less chance that the user will miss them).

**Reply by smichel17** [reply]
I like this overall direction very much. I'm not totally sure if those are the best 3 modes, but that's a minor detail that can be worked out later.

One concern: how does the location services prompt work with this? Right now when you press the gps button but location services are off, you get a prompt to enable them. However, location services aren't a hard requirement to use SC (although they are strongly recommended), and I wouldn't want to take "rotate north" functionality away from someone who wanted to leave gps off (or annoy them with a repetitive dialog).

**Reply by mnalis** [reply]
On (non-cached) app startup, map could start in `fully manual` mode which does not require GPS. Then first press on the `map mode` icon would switch to  `Full 3D` (which requires GPS for `locked camera`, and thus would ask for android location permission).

**Reply by smichel17** [reply]
Right, the thing is that *right now*, if I **never** enable gps and/or stay in fully-manual mode, I can tap the compass to rotate so North is up, and then tap it again to tilt the map. In this case, I would not be able to do that because the app would ask me to turn on location services as soon as I tap the `map mode` icon.

edit: Personally, right now I mostly use manual camera, and I regularly use the compass to orient the map so North is up.

**Reply by mnalis** [reply]
True, that would be inconvenient. I've also just noted that in the hurry I failed to add the mode that I actually use the most :smile: which is `2D no rotate` (eg. 2D map stays in the manually set orientation, but follows the GPS). 

Perhaps to solve this issue with "North is up (without GPS)" there could be another mode, which is just `manual 2D North is up` (eg. 2D, north is up, without any compass and direction) and start in that mode. But I'd be wary of adding too many different modes. 5 might be okay(-ish), but 10 is definitely not :thinking:

**Comment by smichel17** [comment]
> Good point about nonsensical states (although that maybe might be worked around by hiding buttons sometimes, as we do for GPS button at the moment)

> Or, perhaps, replace both compass and GPS icons with just one map mode icon, which would cycle through small number of actually useful modes on each button click (and change how that icon looks):

Another option would be some hybrid of these, where there is a `map mode` button *and* a second button, which modifies some behavior of the current mode. This is something like the inverse of the current approach, where the gps tracking mode modifies the behavior of the compass.

~~The advantage of this approach is that we could *change the icon* of the secondary button based on which mode was active.~~ As I write this, I realize that we could totally do this with the current approach: change the compass icon when the camera is locked. Maybe that's all we need, not a major overhaul of functionality; maybe the real problem all along has been **hidden state**, and making it visible would solve the problem. (i.e. show a different icon in the compass position depending on whether the camera is locked).

Or, here's another option: when the camera is locked, **always follow** the compass orientation (whether the view is tilted or flat/birdseye).

When https://github.com/streetcomplete/StreetComplete/issues/1040 is finished, perhaps birdseye should follow the direction of movement instead. Or both could use some combination of the direction of movement and the compass direction. We can figure it out later. The point is, now the compass behavior would be this:

- When the camera is locked, it only switches between flat and tilted.
- When the camera is unlocked and pointing north, it only switches between flat and tilted.
    - When the camera is unlocked and not pointing north, it rotates to point north and switch to flat.

Or, in other words, **at all times** the camera just switches between flat and tilted, **except** when unlocked and not pointing north, in which case it first rotates to point north.

**Comment by matkoniecz** [comment]
Note that in case that this brainstorming produces some actionable outcome - it would be likely a good idea to make an issue proposing something specific and having a summary of reasoning.

**Reply by smichel17** [reply]
Definitely will do. I've been experimenting recently with erring on the side of starting discussions for anything which is not really concrete, actionable, and likely to be accepted. So far I think it's a good thing; the limited threading ability of discussions is nice for any kind of branching conversation, which these often are. It also helps keep issues actionable instead of sprawling discussion threads like #1654 :sweat_smile:, and it's a good fit with @westnordost's desire to keep open issues below 100.

**Comment by FloEdelmann** [comment]
Let's not focus on this too much, as there will be changes to the buttons and navigation modes anyway soon :slightly_smiling_face: 

https://github.com/streetcomplete/StreetComplete/compare/navigation-mode

https://github.com/streetcomplete/StreetComplete/issues/1040#issuecomment-928062073

**Reply by smichel17** [reply]
Well, this discussion was 4 days ago, right at the time of the comment you linked. So, before the implementation is complete/released, seems to be the *best* time to discuss this. After all, if there's going to be changes anyway, that's the best time to get it right, to keep the number of UI changes to a minimum.

**Comment by smichel17** [comment]
I haven't gone mapping with it, but I experimented with https://github.com/streetcomplete/StreetComplete/pull/3335 to see how it is. I think it **resolves the confusing behavior**. It works similarly to what @mnalis proposed [above](https://github.com/streetcomplete/StreetComplete/discussions/3326#discussioncomment-1405167):

There are three modes: **unlocked**; locked (**top-down**); and locked (**navigation mode**).

The camera **starts unlocked**, and you can enter this mode at any time by **panning** the map.
Tapping the GPS button while **unlocked** will resume the **previous mode** (default: top-down).
Tapping the GPS button while **locked** will **toggle** to the other mode.

The compass disappears when the map is pointing north and at birdseye view. Pressing it orients the map in that way; this does not unlock the camera, but does change the mode to top-down, if you were in navigation mode.

I have no complaints so far; I'll try it at some point and post feedback in https://github.com/streetcomplete/StreetComplete/pull/3335, if it's not merged by then.

**Reply by westnordost** [reply]
The idea for the compass button is that it always only does one thing so that it is not confusing: Orient the map back to North.
When the map is already oriented North, thus the button does not need to be shown and we get [some free screen real estate](https://www.youtube.com/watch?v=yNxPVj0hejg).
